File: cms/backend/src/services/mediaValidationService.ts
import fs from "node:fs";
import path from "node:path";
import { env } from "../config/env.js";
import { mediaLibraryRepository } from "../repositories/jsonRepositories.js";
import { HttpError } from "../utils/http.js";
import { sanitizePath, sanitizeText } from "../utils/sanitize.js";

export type InternalMediaKind = "image" | "video" | "all";

export const IMAGE_EXTENSIONS = new Set([
  ".png",
  ".jpg",
  ".jpeg",
  ".webp",
  ".gif",
  ".svg",
  ".avif",
]);

export const VIDEO_EXTENSIONS = new Set([".mp4", ".webm", ".ogg"]);
export const MEDIA_EXTENSIONS = new Set([...IMAGE_EXTENSIONS, ...VIDEO_EXTENSIONS]);

export interface InternalMediaValidationOptions {
  kind?: InternalMediaKind;
  required?: boolean;
  label?: string;
}

function normalizePublicPath(value: unknown) {
  const raw = sanitizeText(value, 600);
  if (!raw) return "";
  if (/^[a-z][a-z0-9+.-]*:/i.test(raw)) return "";
  const pathValue = sanitizePath(raw.startsWith("/public/") ? raw.slice("/public".length) || "/" : raw);
  return pathValue;
}

function safeFilePathForUrl(url: string) {
  const normalized = normalizePublicPath(url);
  if (!normalized) return "";

  const root = normalized.startsWith("/uploads/")
    ? env.uploadsDir
    : env.frontendPublicDir;
  const relative = normalized.startsWith("/uploads/")
    ? normalized.replace(/^\/uploads\//, "")
    : normalized.replace(/^\//, "");
  const resolved = path.resolve(root, relative);
  const resolvedRoot = path.resolve(root);

  if (resolved !== resolvedRoot && !resolved.startsWith(`${resolvedRoot}${path.sep}`)) {
    return "";
  }

  return resolved;
}

function urlMatchesKind(url: string, kind: InternalMediaKind) {
  const ext = path.extname(url).toLowerCase();
  if (kind === "image") return IMAGE_EXTENSIONS.has(ext);
  if (kind === "video") return VIDEO_EXTENSIONS.has(ext);
  return MEDIA_EXTENSIONS.has(ext);
}

function knownUrlsFromLibrary() {
  const urls = new Set<string>();
  for (const item of mediaLibraryRepository.read()) {
    for (const key of [
      "url",
      "optimizedUrl",
      "originalUrl",
      "thumbnailUrl",
      "smallUrl",
      "mediumUrl",
      "largeUrl",
      "posterUrl",
    ]) {
      const value = normalizePublicPath(item[key]);
      if (value) urls.add(value);
    }
  }
  return urls;
}

export function normalizeInternalMediaUrl(value: unknown): string {
  return normalizePublicPath(value);
}

export function isKnownLibraryMedia(value: unknown, kind: InternalMediaKind = "all"): boolean {
  const url = normalizeInternalMediaUrl(value);
  if (!url || !urlMatchesKind(url, kind)) return false;

  const filePath = safeFilePathForUrl(url);
  try {
    return Boolean(filePath && fs.existsSync(filePath) && fs.statSync(filePath).isFile());
  } catch {
    return false;
  }
}

export function assertInternalMediaUrl(
  value: unknown,
  options: InternalMediaValidationOptions = {}
): string {
  const kind = options.kind ?? "all";
  const label = options.label ?? "Mídia";
  const raw = sanitizeText(value, 600);
  const url = normalizeInternalMediaUrl(value);

  if (!url) {
    if (raw) {
      throw new HttpError(422, `${label}: use somente arquivos internos da biblioteca de mídia.`);
    }
    if (options.required) throw new HttpError(422, `${label}: selecione uma mídia da biblioteca.`);
    return "";
  }

  if (!urlMatchesKind(url, kind)) {
    throw new HttpError(
      422,
      `${label}: tipo de arquivo incompatível com o campo.`
    );
  }

  if (!isKnownLibraryMedia(url, kind)) {
    throw new HttpError(
      422,
      `${label}: use somente arquivos internos da biblioteca de mídia.`
    );
  }

  return url;
}

export function sanitizeInternalImageUrl(value: unknown, label = "Imagem"): string {
  return assertInternalMediaUrl(value, { kind: "image", label });
}

export function sanitizeInternalVideoUrl(value: unknown, label = "Vídeo"): string {
  return assertInternalMediaUrl(value, { kind: "video", label });
}

export function sanitizeInternalMediaUrl(value: unknown, label = "Mídia"): string {
  return assertInternalMediaUrl(value, { kind: "all", label });
}

File: cms/backend/src/main/java/br/com/rodogarcia/cms/backend/service/MediaValidationService.java
package br.com.rodogarcia.cms.backend.service;

import java.nio.file.Files;
import java.nio.file.InvalidPathException;
import java.nio.file.Path;
import java.util.Locale;
import java.util.Set;

import br.com.rodogarcia.cms.backend.config.CmsProperties;
import br.com.rodogarcia.cms.backend.exception.ApiException;
import br.com.rodogarcia.cms.backend.model.content.ContentJson;
import br.com.rodogarcia.cms.backend.utils.Sanitizers;
import org.springframework.stereotype.Service;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.node.StringNode;

@Service
public final class MediaValidationService {
    public enum Kind { IMAGE, VIDEO, ALL }

    public static final Set<String> IMAGE_EXTENSIONS = Set.of(
        ".png", ".jpg", ".jpeg", ".webp", ".gif", ".svg", ".avif"
    );
    public static final Set<String> VIDEO_EXTENSIONS = Set.of(".mp4", ".webm", ".ogg");
    public static final Set<String> MEDIA_EXTENSIONS = Set.of(
        ".png", ".jpg", ".jpeg", ".webp", ".gif", ".svg", ".avif",
        ".mp4", ".webm", ".ogg"
    );

    private final Path uploadsRoot;
    private final Path publicRoot;

    public MediaValidationService(CmsProperties properties) {
        this.uploadsRoot = properties.uploadsDir().toAbsolutePath().normalize();
        this.publicRoot = properties.frontendPublicDir().toAbsolutePath().normalize();
    }

    public String assertInternal(JsonNode value, Kind kind, boolean required, String label) {
        String raw = Sanitizers.text(value, 600);
        String url = normalize(value);
        if (url.isEmpty()) {
            if (!raw.isEmpty()) {
                throw new ApiException(422, label + ": use somente arquivos internos da biblioteca de mídia.");
            }
            if (required) {
                throw new ApiException(422, label + ": selecione uma mídia da biblioteca.");
            }
            return "";
        }
        if (!matches(url, kind)) {
            throw new ApiException(422, label + ": tipo de arquivo incompatível com o campo.");
        }
        if (!isKnown(url, kind)) {
            throw new ApiException(422, label + ": use somente arquivos internos da biblioteca de mídia.");
        }
        return url;
    }

    public String assertInternal(String value, Kind kind, boolean required, String label) {
        return assertInternal(StringNode.valueOf(value == null ? "" : value), kind, required, label);
    }

    public String normalize(JsonNode value) {
        String raw = Sanitizers.text(value, 600);
        if (raw.isEmpty() || ContentJson.hasScheme(raw)) return "";
        String candidate = raw.startsWith("/public/")
            ? raw.substring("/public".length())
            : raw;
        return Sanitizers.path(StringNode.valueOf(candidate));
    }

    public String normalize(String value) {
        return normalize(StringNode.valueOf(value == null ? "" : value));
    }

    public boolean isKnown(String raw, Kind kind) {
        String url = normalize(raw);
        if (url.isEmpty() || !matches(url, kind)) return false;
        try {
            Path root = url.startsWith("/uploads/") ? uploadsRoot : publicRoot;
            String relative = url.startsWith("/uploads/")
                ? url.substring("/uploads/".length())
                : url.substring(1);
            Path resolved = root.resolve(relative).toAbsolutePath().normalize();
            return !resolved.equals(root) && resolved.startsWith(root) && Files.isRegularFile(resolved);
        } catch (InvalidPathException | SecurityException ignored) {
            return false;
        }
    }

    public static boolean matches(String url, Kind kind) {
        String extension = extension(url);
        return switch (kind) {
            case IMAGE -> IMAGE_EXTENSIONS.contains(extension);
            case VIDEO -> VIDEO_EXTENSIONS.contains(extension);
            case ALL -> MEDIA_EXTENSIONS.contains(extension);
        };
    }

    public static String extension(String value) {
        int slash = value.lastIndexOf('/');
        int dot = value.lastIndexOf('.');
        return dot > slash ? value.substring(dot).toLowerCase(Locale.ROOT) : "";
    }
}
